sdk/opa.go

```diff
@@ -0,0 +1,405 @@
+			return err
+		}
+
+		record := &server.Info{
+			Txn:        txn,
+			Bundles:    bundles,
+			DecisionID: result.ID,
+			Timestamp:  now,
+			Path:       path,
+			Input:      &params.Input,
+			InputAST:   inputAST,
+			Error:      err,
+			Metrics:    m,
+		}
+
+		if result.Result != nil {
+			record.Results = &result.Result
+		}
+
+		m.Timer(metrics.SDKDecisionEval).Stop()
+
+		if err := logger.Log(ctx, record); err != nil {
+			return fmt.Errorf("decision log: %w", err)
+		}
+
+		return err
+	})
+
+	return result, err
+}
+
+func (opa *OPA) evaluate(ctx context.Context, txn storage.Transaction, now time.Time, path string, input interface{}, m metrics.Metrics) (interface{}, ast.Value, map[string]server.BundleInfo, error) {
+
+	bundles, err := opa.bundles(ctx, txn)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	r, err := ref.ParseDataPath(path)
+	if err != nil {
+		return nil, nil, bundles, err
+	}
+
+	pq, err := opa.prepare(ctx, txn, now, r.String(), m)
+	if err != nil {
+		return nil, nil, bundles, err
+	}
+
+	inputAST, err := ast.InterfaceToValue(input)
+	if err != nil {
+		return nil, nil, bundles, err
+	}
+
+	rs, err := pq.Eval(
+		ctx,
+		rego.EvalTime(now),
+		rego.EvalParsedInput(inputAST),
+		rego.EvalTransaction(txn),
+		rego.EvalMetrics(m),
+		rego.EvalInterQueryBuiltinCache(opa.interQueryBuiltinCache),
+	)
+	if err != nil {
+		return nil, inputAST, bundles, err
+	} else if len(rs) == 0 {
+		return nil, inputAST, bundles, undefinedDecisionErr(path)
+	}
+
+	return rs[0].Expressions[0].Value, inputAST, bundles, nil
+}
+
+func (opa *OPA) pluginStatusUpdated(status map[string]*plugins.Status) {
+	opa.mtx.Lock()
+	defer opa.mtx.Unlock()
+
+	if opa.ready {
+		return
+	}
+
+	// NOTE(tsandall): we do not include a special case for the discovery
+	// plugin. If the discovery plugin is the only plugin and it goes ready,
+	// then OPA will be considered ready. The discovery plugin only goes ready
+	// _after_ it has successfully processed a discovery bundle. During
+	// discovery bundle processing, other plugins will register so their states
+	// will be accounted for. If a discovery bundle did not enable any other
+	// plugins (bundles, etc.) the OPA will still be operational.
+
+	for _, s := range status {
+		if s.State != plugins.StateOK {
+			return
+		}
+	}
+
+	close(opa.readyCh)
+	opa.ready = true
+}
+
+func (opa *OPA) compilerUpdated(_ storage.Transaction) {
+	opa.mtx.Lock()
+	defer opa.mtx.Unlock()
+	opa.preparedQueries = map[string]*rego.PreparedEvalQuery{}
+}
+
+func (opa *OPA) prepare(ctx context.Context, txn storage.Transaction, now time.Time, query string, m metrics.Metrics) (*rego.PreparedEvalQuery, error) {
+	opa.mtx.Lock()
+	defer opa.mtx.Unlock()
+
+	result, ok := opa.preparedQueries[query]
+	if !ok {
+		pq, err := rego.New(
+			rego.Time(now),
+			rego.Metrics(m),
+			rego.Query(query),
+			rego.Compiler(opa.manager.GetCompiler()),
+			rego.Store(opa.manager.Store),
+			rego.Transaction(txn),
```

Comment:
What does it mean that we hold a reference to a transaction (and I guess other per-query specific references like metrics) given that this query will be cached for further use? The transaction stored here won't be the one used in any other queries except for the first.. so I'm guessing this is unused?

Reply:
Just confusing to me that we store stale snapshot data like the transaction and the query metrics on a `rego` object which is reused for each query. If I looked at the object in a debugger I would assume that those values were there because they were needed/used.

> If we don't provide one, the rego package will open one anyway. There's no point in opening two so just provide it here.

I don't get it :) If the transaction stored is stale, won't the rego package have to open one anyway?